Various tooling improvements from other plugins

lieut-data opened the pull request:
#### Summary
Upstream various changes from other plugins:
- collapse `golangci-lint` into just the existing `check-style` target
- support `DEBUG=1 server` target
- eliminate `(debug|deploy)-dist` to `DEBUG=1 (dist|deploy)` for simplicity
- introduce `attach` and `attach-headlesss` for easy `dlv` setup
- `test` shouldn't change files via `npm run fix`
- introduce `reset` to disable/enable plugin
- introduce `kill` to forcibly kill all instances of a plugin
- rename `build/deploy` to `build/pluginctl`, now with `reset` functionality
- eliminate copy-to-install behaviour from `build/pluginctl`
- enable `--cache` when running eslint
- `.npmrc` specifies `save-exact=true`
- introduce `watch` to build, deploy and update webapp changes automatically
- touch `node_modules` instead of `.npminstall`
- support [`devtoolNamespace`](https://webpack.js.org/configuration/output/#outputdevtoolnamespace)

Fixes: https://mattermost.atlassian.net/browse/MM-11867

lieut-data commented:
Skipping `3: QA Review`, since only tooling changes.

lieut-data commented:
@cpoile, my hope is to downstream this back to incident response, with just one additional change in `build/custom.mk`, otherwise having `Makefile` be 100% a match for the starter-template.

I took the opportunity to slim out some of the code in incident-response around copying plugins instead of Client4-installing them, since GOPATH is long dead, and the `mmctl` strategy is fast becoming the popular approach. I also took some opinionated strategies with target naming, also removing the background nature of `attach-headless` and forcing you to `Ctrl-C` it instead.

hanzei commented:
@lieut-data is there any urgency with reviewing this PR? I would love to understand the details before giving a review, but this will take time.

lieut-data commented:
@hanzei, there's no immediate hurry -- I can propose the same changes into incident response in parallel, and if we diverge, even up later.

cpoile commented on Makefile:
```diff
@@ -60,6 +61,16 @@ ifneq ($(HAS_SERVER),)
 endif
 
+## Builds the server, if it exists, in debug mode for all supported architectures.
+.PHONY: server-debug
+server-debug:
+ifneq ($(HAS_SERVER),)
+	mkdir -p server/dist
+	cd server && env GOOS=darwin GOARCH=amd64 $(GO) build $(GO_BUILD_FLAGS) -gcflags "all=-N -l" -o dist/plugin-darwin-amd64;
+	cd server && env GOOS=linux GOARCH=amd64 $(GO) build $(GO_BUILD_FLAGS) -gcflags "all=-N -l" -o dist/plugin-linux-amd64;
+	cd server && env GOOS=windows GOARCH=amd64 $(GO) build $(GO_BUILD_FLAGS) -gcflags "all=-N -l" -o dist/plugin-windows-amd64.exe;
+endif
```
Just wondering if there's a reason for removing the single-arch build logic?

lieut-data replied:
I found the behaviour "surprising", in that it assumed I was building and deploying to a debug version to the same machine / architecture. I realize that's intently common for Mattermost stuff, but wondering if there is a potential use case for not doing that?

I had half a mind to formalize an architecture override -- for both production and debug builds -- with an environment variable or some such. Thoughts?

cpoile replied:
Maybe it's more useful in the way server-debug works in IR: build just for your arch, then upload just the server side of the plugin, then reset the plugin. It takes a second or two, so it's useful when developing. I guess we're removing that functionality here?

lieut-data replied:
Ah, that makes sense. Let me revisit all of this re: watching.

lieut-data replied:
I ended up leaving the build-all-architectures here for the time being, despite restoring `make watch` as described below. Let me know if you think this is a blocker!

cpoile commented on Makefile:
```diff
@@ -74,7 +85,7 @@ ifneq ($(HAS_WEBAPP),)
 
-## Builds the webapp in debug mode, if it exists.
+## Builds the webapp, in debug mode, if it exists.
 .PHONY: webapp-debug
```
Would you be willing to move over the webapp-debug logic from incident response also? I find it useful almost everyday when working on the frontend.

lieut-data replied:
Yes! I realized after digging in a bit that I had missed part of your improvements. This is also a good note to myself to update README.md with something to help guide developers starting with this template.

lieut-data replied:
Moved! I reworked it slightly to use `pluginctl` to redeploy (with a quick `bundle` vs. having to rebuild everything) so that we don't have to assume `mattermost-server` is a peer directory at all. It's now just `make watch` to get this functionality.

mickmister commented:
I apparently missed this PR and have a similar effort at https://github.com/mattermost/mattermost-plugin-starter-template/pull/99 . I got most of the inspiration from the Incident Response plugin as it has a pretty robust dev setup. Two additions I've made:

- [`make webapp-debug-deploy`](https://github.com/mattermost/mattermost-plugin-starter-template/blob/b1deec513f92594602d6a01a6f04454516ffcf72/Makefile#L266): Compile webpack with dev config settings and deploy the plugin, but do not use watch mode, as opposed to `webapp-debug` which does use watch mode.
- Add [`devtoolNamespace`](https://github.com/mattermost/mattermost-plugin-starter-template/blob/b1deec513f92594602d6a01a6f04454516ffcf72/webapp/webpack.config.js#L83) to the webpack config to allow the plugin to have its own namespace in the dev tools source tab. Docs can be found [here](https://webpack.js.org/configuration/output/#outputdevtoolnamespace). Here's an example with the Jira and GitHub plugins:

![image](https://user-images.githubusercontent.com/6913320/83480030-41759380-a468-11ea-8a25-dde54ff2b191.png)

I mainly made the PR to get some eyes on it before holistically applying the additions to the other plugin repos. It would be immensely helpful to get all of the plugins up to speed on these configurations!

I know there is a task to have various aspects of the repos automatically synchronized, but I think this first go-through should be done manually as there are some intricacies to make sure it is all configured correctly.

mickmister commented:
Another good thing to add would be an [.npmrc](https://github.com/mattermost/mattermost-mobile/blob/master/.npmrc) file at the root of the webapp folder, that enforces dependencies to be installed with `save-exact`.

lieut-data commented:
@mickmister, I've gone ahead and made the following changes:
* Added the requisite `.npmrc`
* In lieu of `make webapp-debug-deploy`, I've just made `DEBUG=1 deploy`, with `make watch` (or `DEBUG=1 make watch`) for the specialized watching behaviour. Note that I moved away from the `-debug` variants since there were so many desired permutations, and it seemed easier to just export `DEBUG=1` into my environment and get this "for free".
* Added devtoolNamespace as per your excellent recommendation

Can you and @cpoile give this a whirl and see if it fits your workflows / make intuitive sense?

mickmister commented:
@lieut-data I'm not sure I see much benefit from using `make watch` without `DEBUG=1`, so it seems off to me to have non-debug to be the default. I don't see why I would want the watch behavior, without the benefits of the debug build, but I suppose I understand if it is desirable to be able to replicate the production build in rapid development.

Regardless of that point, I do like the explicit mention of `watch` in the command.

lieut-data commented:
> I suppose I understand if it is desirable to be able to replicate the production build in rapid development

Yeah, this was exactly my imagined use case. I've often had to debug issues that didn't manifest in debug builds, but could be reproduce in production builds. By exporting `DEBUG=1` to your environment, you could get the desired default in most cases. Actually, this reminds me that I haven't updated the `README.md` to document all this -- will do before merging.

mickmister commented:
I think a `make disable` command would be great to include here. It would do the same thing `make reset` does, but without the final enable. I've just had the need for one and it seems natural to use.

mickmister commented on webapp/webpack.config.js:
```diff
@@ -36,8 +66,12 @@ module.exports = {
         'react-router-dom': 'ReactRouterDom',
     },
     output: {
+        devtoolNamespace: PLUGIN_ID,
```
I think we should add some other loaders such as the scss loader. Adding dev dependencies like that doesn't increase bundle size, so there's no bloat from adding it.

`npm install --save-dev style-loader css-loader sass-loader`

Then in the `webpack.config.js` [rules](https://github.com/mattermost/mattermost-plugin-jira/blob/dd988f94ebcf834806aeabefc470b410977dc187/webapp/webpack.config.js#L48-L64):

```js
{
    test: /\.scss$/,
    use: [
        'style-loader',
        {
            loader: 'css-loader',
        },
        {
            loader: 'sass-loader',
            options: {
                includePaths: ['node_modules/compass-mixins/lib', 'sass'],
            },
        },
    ],
}
```

lieut-data replied:
Great suggestions! Updated.

mickmister commented:
I'd also like to mention an open PR that is

lieut-data commented:
Thanks, @mickmister! I've updated with your excellent suggestions.
